lightning-c-bindings/src/c_types/mod.rs

```diff
@@ -0,0 +1,415 @@
+impl PublicKey {
+	pub(crate) fn from_rust(pk: &SecpPublicKey) -> Self {
+		Self {
+			compressed_form: pk.serialize(),
+		}
+	}
+	pub(crate) fn into_rust(&self) -> SecpPublicKey {
+		SecpPublicKey::from_slice(&self.compressed_form).unwrap()
+	}
+	pub(crate) fn is_null(&self) -> bool { self.compressed_form[..] == [0; 33][..] }
+	pub(crate) fn empty() -> Self { Self { compressed_form: [0; 33] } }
+}
+
+#[repr(C)]
+pub struct SecretKey {
+	pub bytes: [u8; 32],
+}
+impl SecretKey {
+	// from_rust isn't implemented for a ref since we just return byte array refs directly
+	pub(crate) fn from_rust(sk: SecpSecretKey) -> Self {
+		let mut bytes = [0; 32];
+		bytes.copy_from_slice(&sk[..]);
+		Self { bytes }
+	}
+	pub(crate) fn into_rust(&self) -> SecpSecretKey {
+		SecpSecretKey::from_slice(&self.bytes).unwrap()
+	}
+}
+
+#[repr(C)]
+pub struct Signature {
+	pub compact_form: [u8; 64],
+}
+impl Signature {
+	pub(crate) fn from_rust(pk: &SecpSignature) -> Self {
+		Self {
+			compact_form: pk.serialize_compact(),
+		}
+	}
+	pub(crate) fn into_rust(&self) -> SecpSignature {
+		SecpSignature::from_compact(&self.compact_form).unwrap()
+	}
+	pub(crate) fn is_null(&self) -> bool { self.compact_form[..] == [0; 64][..] }
+	pub(crate) fn empty() -> Self { Self { compact_form: [0; 64] } }
+}
+
+#[repr(C)]
+/// A reference to a serialized transaction, in (pointer, length) form.
+/// This type does *not* own its own memory, so access to it after, eg, the call in which it was
+/// provided to you are invalid.
+pub struct Transaction {
+	pub data: *const u8,
+	pub datalen: usize,
+}
+impl Transaction {
+	pub(crate) fn into_bitcoin(&self) -> BitcoinTransaction {
+		// datalen == 0 is invalid, so don't bother checking here.
+		::bitcoin::consensus::encode::deserialize(unsafe { std::slice::from_raw_parts(self.data, self.datalen) }).unwrap()
+	}
+	pub(crate) fn from_slice(s: &[u8]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+}
+
+#[repr(C)]
+#[derive(Clone)]
+/// A transaction output including a scriptPubKey and value.
+/// This type *does* own its own memory, so must be free'd appropriately.
+pub struct TxOut {
+	pub script_pubkey: derived::CVec_u8Z,
+	pub value: u64,
+}
+
+impl TxOut {
+	pub(crate) fn into_rust(self) -> ::bitcoin::blockdata::transaction::TxOut {
+		::bitcoin::blockdata::transaction::TxOut {
+			script_pubkey: self.script_pubkey.into_rust().into(),
+			value: self.value,
+		}
+	}
+	pub(crate) fn from_rust(txout: ::bitcoin::blockdata::transaction::TxOut) -> Self {
+		Self {
+			script_pubkey: CVecTempl::from(txout.script_pubkey.into_bytes()),
+			value: txout.value
+		}
+	}
+}
+#[no_mangle]
+pub extern "C" fn TxOut_free(_res: TxOut) { }
+
+#[repr(C)]
+pub struct u8slice {
+	pub data: *const u8,
+	pub datalen: usize
+}
+impl u8slice {
+	pub(crate) fn from_slice(s: &[u8]) -> Self {
+		Self {
+			data: s.as_ptr(),
+			datalen: s.len(),
+		}
+	}
+	pub(crate) fn to_slice(&self) -> &[u8] {
+		if self.datalen == 0 { return &[]; }
```

Comment:
this isn't needed if `data` is never null.
(maybe document how each type can/should be constructed/modified, I'll guess some are only allowed to be constructed on rust and some also in C/C++)

Reply:
In general, pretty much anything can be constructed in C/C++, AFAIU. Because Rust cdylib is guaranteed to use malloc()/free() for everything, as long as the C end does the same, Rust will free it appropriately. eg the demo.cpp app does it for a LDKC2Tuple_u64u64Z, allocating two u64s and letting Rust free them.

Reply:
That's implementation defined and can be changed by any library in the tree (and in libstd in the future, for example in the past rust used jemalloc instead)
I wouldn't rely on rust to free C pointers and vise versa.

Reply:
Wasnt jemalloc only for non-cdylib things? Certainly we could pull in a dependency that does something insane, but my understanding was malloc()/free() was guaranteed for cdylib-builds as long as nothing in your tree does a custom global allocator.

Reply:
Ha, it seems like you are right: https://doc.rust-lang.org/std/alloc/index.html
> Currently the default global allocator is unspecified. Libraries, however, like cdylibs and staticlibs are guaranteed to use the System by default.

Reply:
FWIW, on windows `HeapAlloc` and `HeapFree` are used by rust instead of malloc/free, idk if the windows libc also uses those for the malloc/free implementation or not
https://github.com/rust-lang/rust/blob/master/library/std/src/sys/windows/alloc.rs#L44

Reply:
Yeesh wtf. Can we just not support Windowz?

Reply:
I added a note in the README pointing to the System alloc docs to describe which allocator must be used.